### PR TITLE
Fix function body for NLL Loss and SCE Loss

### PR DESCRIPTION
**Pull request by ashbhandare**
The function body of NLL Loss assumes a float input type and errors out for MLFloat16 currently.
The SCE Loss function body computes the log_prob with an unstable result(possibly due to the Log op), which this PR fixes by using LogSoftmax instead resulting in a more stable result.
This PR also fixes a bug in test generation.

### CODE REVIEW

File: onnx/defs/math/defs.cc [hunk 1]
```diff
@@ -3412,19 +3438,47 @@ bool BuildContextDependentFunctionBodySCE(
...
 
+  // Using stable implementation of LogSoftmax
+  // We need to unsqueeze the inputs on the last axis if they are 2D in order to make the 
+  // further Reshape and transpose valid.
+  body.push_back(
+      {{"X_unsqueezed"},
+        "Unsqueeze",
+       {"scores", "axes"}});
+  body.push_back(
+      {{"3D_shape"},
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:3451]: Suggest renaming "3D_shape" to a valid C identifier, eg., "ThreeD_shape" or "Shape3D".

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -3412,19 +3438,47 @@ bool BuildContextDependentFunctionBodySCE(
       {{"axes"},
        "Constant",
        {},
-       {MakeAttribute("value", ToDimensionOneInt64Tensor(1))}});
+       {MakeAttribute("value", ToDimensionOneInt64Tensor(-1))}});
 
+  // Using stable implementation of LogSoftmax
+  // We need to unsqueeze the inputs on the last axis if they are 2D in order to make the 
+  // further Reshape and transpose valid.
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:3445]: Just curious: is this really required? It seems like reshaping a 2D with target-shape {0,0,-1} should have the same effect, if I look at the reshape spec. Wonder if there is some bug/ambiguity in Reshape impl/spec.

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -2847,6 +2853,13 @@ bool BuildContextDependentFunctionBody(
     const FunctionBodyBuildContext& ctx,
     const OpSchema& schema,
     FunctionProto& functionProto) {
+  bool float_input = true;
+  if (ctx.getInputType(0) != nullptr) {
+    float_input = ctx.getInputType(0)->tensor_type().elem_type() == TensorProto_DataType_FLOAT;
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:2858]: minor nit: ```ctx.getInputType(0)->tensor_type().elem_type()``` is repeated thrice. May be worth making it a variable. Changing the control-flow to ```if (ctx.getInputType(0) == nullptr) return;``` would make it doable.

File: onnx/defs/math/defs.cc (hunk 2 above)
Comment (ashbhandare) [reply onnx/defs/math/defs.cc:3445]: you are right, this unsqueeze can be skipped

**Review by gramalingam — APPROVED** [review]

**Review by askhade — APPROVED** [review]

**Review by postrational — APPROVED** [review]

**Pull request merged**